pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,660 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/blang/semver/v4"
+	v1 "k8s.io/api/coordination/v1"
+	v1alpha1 "k8s.io/api/coordination/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coordinationv1informers "k8s.io/client-go/informers/coordination/v1"
+	coordinationv1alpha1 "k8s.io/client-go/informers/coordination/v1alpha1"
+	coordinationv1client "k8s.io/client-go/kubernetes/typed/coordination/v1"
+	coordinationv1alpha1client "k8s.io/client-go/kubernetes/typed/coordination/v1alpha1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+const (
+	controllerName          = "leader-election-controller"
+	ElectedByAnnotationName = "coordination.k8s.io/elected-by" // Value should be set to controllerName
```

Comment:
is this an API ?

Reply:
It's a little optimization for us to recognize it when a holder understands CLE and PreferredHolder. I'm trying to only operate on CLE aware Leases and expired Leases.